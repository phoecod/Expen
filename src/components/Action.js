import React from 'react';
import ReactModal from 'react-modal';

ReactModal.setAppElement('#app');

const action = class Action extends React.Component {
	
	constructor (props) {
		super(props);
		this.state = {
			selectedOption: undefined,
			showModal: false
		}
	}

	chooseOption = () => {
		const option = this.props.options[Math.floor(Math.random()*this.props.options.length)];
		if (!!option) {
			this.setState(() => {
				return {
					selectedOption: option,
					showModal: true
				}
			}); 
		}
	}

	handleCloseModal  = () => {
		this.setState({ showModal: false});
	}

	render = () => {
		const optionsLength = this.props.options.length > 0;
		return (
			<div>
				<button className="big-button" disabled={!optionsLength} onClick={this.chooseOption}>what should I do?</button>
				<ReactModal
					className = "modal"
					isOpen={this.state.showModal}
					onRequestClose={this.handleCloseModal}
					contentLabel="Selected Option"
					className="modal"
					closeTimeoutMS={500}
				>
					<h1 className=".modal-title">Selected Option</h1>
					<h2 className="modal-body">{this.state.selectedOption}</h2>
					<button className="button" onClick={this.handleCloseModal}>Back</button>
				</ReactModal>
			</div>
		);
	}
}

export default action;
