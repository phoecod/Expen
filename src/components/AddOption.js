import React from 'react';
import { CSSTransitionGroup } from 'react-transition-group';

export default class AddOption extends React.Component {

  state = {
    error: undefined
  }
  
  addToOptions = (e) => {    
    e.preventDefault();
    const value = e.target.elements.option.value.trim();
    const errorMsg = this.props.handleAddOption(value);
    this.setState(() => {
      return {
        error: errorMsg
      }
    });
    if (!errorMsg) {
      e.target.elements.option.value = '';
    }
  }
  render () {
    return (
      <div className="">
        {this.state.error && <p className="add-option-error">{this.state.error}</p>}
        <form className="add-option" onSubmit={this.addToOptions}>
          <input className="add-option-input" type="text" name="option"></input>
          <button className="button">Add Option</button>
        </form>
      </div>
      
    );
  }
}