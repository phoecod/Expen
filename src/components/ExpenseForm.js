import React from 'react';
import moment from 'moment';
import {connect} from 'react-redux';
import  { SingleDatePicker } from 'react-dates';



export const ExpenseForm = class ExpenseForm extends React.Component {

	constructor (props) {
		super (props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? props.expense.amount.toString() : '',
			note: props.expense ? props.expense.note : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			descError: undefined, 
			amountError: undefined
		}
	}

	handleAddFormExpense = (e) => {
		e.preventDefault();
		const expense = {
			description: this.state.description,
			note: this.state.note,
			amount: parseFloat(this.state.amount, 10),
			createdAt: this.state.createdAt.valueOf()
		}
		if (!this.state.amount) {
			this.setState( () => ({amountError: 'Amount cannot be empty and needs to be greater than 0 with no more than 2 decimal places'}));
			this.setState( () => ({amount: ''}))
		}

		if (this.state.description.length == 0) {
			this.setState(() => ({descError: 'Description cannot be Empty'}));
			this.setState(() => ({description: ''}));
		}

		if (!this.state.descError || !this.state.amountError) {
			this.props.onSubmit(expense);
		}
	}

	handleDescription = (e) => {
		let description = e.target.value.trim();
		if (description.length > 0 || !description ) {
			let descError = undefined;
			this.setState(() => ({description, descError}));
		}
	}

	handleAmount = (e) => {
		let amount = e.target.value.trim();
		if (amount.match(/^\d{1,}(\.\d{0,2})?$/) || !amount) {
			let amountError = undefined;
			this.setState(() => ({amount, amountError}));
		}
	}

	handleNote = (e) => {
		let note = e.target.value.trim();
		this.setState(() => ({note}));
	}

	onDateChange = (createdAt) => {
		this.setState(() => ( {createdAt} ));
	}

	onFocusChange = ({focused}) => {
		this.setState(() => ( {calendarFocused: focused} ));
	}
	
	formPh = {
		description: "Description",
		amount: "Amount",
		note: "Note"
	}

	render () {
		return ( 	
			<div>
				<form onSubmit={this.handleAddFormExpense}>
					{this.state.descError && <p>{this.state.descError}</p>}
					<input value={this.state.description} onChange={this.handleDescription}  placeholder={this.formPh.description} name={this.formPh.description}/><br/>
					{this.state.amountError && <p>{this.state.amountError}</p>}
					<input value={this.state.amount} onChange={this.handleAmount} 
						placeholder={this.formPh.amount} name={this.formPh.amount}/><br/>
					<SingleDatePicker
			            date={this.state.createdAt}
			            onDateChange={this.onDateChange}
			            focused={this.state.calendarFocused}
			            onFocusChange={this.onFocusChange}
			            numberOfMonths={1}
			            isOutsideRange={() => false}
		          	/>
					<textarea 	value={this.state.note}  onChange={this.handleNote} type="text" placeholder={this.formPh.note} name={this.formPh.note}/><br/>
					<button>{this.props.expense ? 'Edit Expense' : 'Add Expense'}</button>
				</form>
			</div>
		);
	}
	
}

const connectedExpenseForm = connect ()(ExpenseForm);

export default connectedExpenseForm;