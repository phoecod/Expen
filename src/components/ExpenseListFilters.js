import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter, setSortByFilter, setStartDateFilter, setEndDateFilter} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export const ExpenseListFilters = class ExpenseListFilters extends React.Component {

	constructor (props) {
		super(props);
		this.state = {
			calendarFocused: null
		}
	}

	onDatesChange = ({startDate, endDate}) => {
		this.props.setStartDateFilter(startDate);
		this.props.setEndDateFilter(endDate);
	}

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({calendarFocused}));
	}

	onTextChange = (e) => {
		const text = e.target.value;
		this.props.setTextFilter(text) ;
	}

	onSelectChange = (e) =>{
		this.props.setSortFilter(e.target.value);
	}

	render () {
		return (
			<div>
				<input id="try" onChange={(e) => this.onTextChange(e)} type="text" value={this.props.filters.text} />
				<select onChange={(e) => this.onSelectChange(e)}>
					<option value="amount">Amount</option>
					<option value="date">Date</option>
				</select>
				<div>
					<DateRangePicker 
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={2}
					isOutsideRange={() => false}
					showClearDates={true}
					/>
				</div>
			</div>
		);
	}
	
}

const stateToProps = (state) => {
	return {
		filters: state.filters
	}
}

const mapDispatchtoProps = (dispatch) => ({
	setStartDateFilter: (startDate) => dispatch(setStartDateFilter(startDate)),
	setEndDateFilter: (endDate) => dispatch(setEndDateFilter(endDate)),
	setSortFilter: (e) => dispatch(setSortByFilter(e)),
	setTextFilter: (text) => dispatch(setTextFilter(text))
});

export default connect(stateToProps,mapDispatchtoProps)(ExpenseListFilters);