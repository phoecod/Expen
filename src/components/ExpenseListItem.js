import React from 'react';
import {startRemoveExpense} from '../actions/expenses';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';




export class ExpenseListItem extends React.Component {

	onClick = () => {
		const itemId = this.props.id;
		this.props.startRemoveExpense(itemId);
	}

	render () {
		return (
			<div>
				<h4> {this.props.description}</h4> 
				<div>
					Amount: {numeral(this.props.amount).format('$0, 0.00')} 
					{this.props.createdAt && <div> Created At: {moment(this.props.createdAt).format("MMM Do YYYY")}</div>}
					{this.props.note && <div>{this.props.note}</div>}
					<div>
						<button className="delete" onClick={ this.onClick }>Delete</button>
						<Link to={`/edit/${this.props.id}`}>
							<button>Edit</button>
						</Link>
					</div>
				</div>
			</div>
		);
	}
	
}

const mapDispatchToProps = (dispatch) => {
	return {
		startRemoveExpense: (id) =>  dispatch(startRemoveExpense({id}))
	}
}

export default connect (undefined, mapDispatchToProps)(ExpenseListItem);