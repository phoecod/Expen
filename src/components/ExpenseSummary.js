import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expensesTotal';
import numeral from 'numeral';

export const ExpenseSummary = ({expensesCount, expensesTotal}) => {

	const expenseWord = expensesCount > 1 ? 'expenses' : 'expense';
	const totalExpensesFormat = numeral(expensesTotal).format('$0.00');
	return (
		<div>
				{
					<h1> 
						Viewing {expensesCount} {expenseWord} totalling {totalExpensesFormat}
					</h1>
				}
		</div>
	);
}

const mapStateToProps = (state) => {
	const visibleExpenses = selectExpenses(state.expenses, state.filters);
	return {
		expensesTotal: selectExpensesTotal(visibleExpenses),
		expensesCount: visibleExpenses.length
	}
}

export default connect(mapStateToProps)(ExpenseSummary);
