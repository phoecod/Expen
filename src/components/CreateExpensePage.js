import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class CreateExpensePage extends React.Component {

	onSubmit = (expense) => {
		this.props.addExpense(expense);
		this.props.history.push('/');
	}
	render() {
		return (
			<div>
				<h3>Add Expense</h3>
				<ExpenseForm onSubmit={this.onSubmit} />
			</div>
		)
	}
}
const mapDispatchtoProps = (dispatch) => ({
	addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchtoProps)(CreateExpensePage);