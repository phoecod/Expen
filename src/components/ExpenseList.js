import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import expenses from '../test/fixtures/expenseData.js';

export const ExpenseList = (props) => {

	return (
		<div>
			{
				props.expenses.length === 0 ? (
					<p> no expenses </p>
				) : (
				props.expenses.map((expense, index) => {
					return <ExpenseListItem key={index} {...expense}/>
				})
				)
			}
		</div>
	);
}
const connectedExpenseList = connect ((state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	}
})(ExpenseList);

export default connectedExpenseList;