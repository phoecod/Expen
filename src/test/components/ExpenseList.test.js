import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenseData from '../fixtures/expenseData';

test('should render ExpenseList with expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={expenseData}/>);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
	const wrapper = shallow(<ExpenseList expenses={[]}/>);
	expect(wrapper).toMatchSnapshot();
});