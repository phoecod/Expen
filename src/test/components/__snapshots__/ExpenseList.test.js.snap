// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
     no expenses 
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="0"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={10}
    createdAt={259200000}
    description="avocado"
    id="3"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={5000}
    createdAt={-1036800000}
    description="Credit Card"
    id="4"
    key="2"
    note=""
  />
</div>
`;
